<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Annokey : Annotation of gene lists with keyword hits from the NCBI gene database and PubMed abstracts" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Annokey</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/annokey">View on GitHub</a>

          <h1 id="project_title">Annokey</h1>
          <h2 id="project_tagline">Annotation of gene lists with keyword hits from the NCBI gene database and PubMed abstracts</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/annokey/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/annokey/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction</h1>

<p>Annokey is a command line tool for annotating gene lists with the results of a keyword (or key phrase) search of the <a href="http://www.ncbi.nlm.nih.gov/gene">NCBI Gene database</a> and linked <a href="http://www.ncbi.nlm.nih.gov/pubmed">PubMed</a> article abstracts. Its purpose is to help users prioritise genes  by relevance to a domain of interest, such as "breast cancer" or "DNA repair" etc. The user steers the search by specifying a ranked list of keywords and terms that are likely to be highly correlated with their domain of interest.</p>

<h3>
<a name="license" class="anchor" href="#license"><span class="octicon octicon-link"></span></a>License</h3>

<p>Annokey is open source software, released under the <a href="https://github.com/bjpop/annokey/blob/master/LICENSE.txt">3-clause BSD license</a>.</p>

<h1>
<a name="features" class="anchor" href="#features"><span class="octicon octicon-link"></span></a>Features</h1>

<p>Annokey's main features are:</p>

<h6>
<a name="online-and-offline-search" class="anchor" href="#online-and-offline-search"><span class="octicon octicon-link"></span></a>Online and offline search</h6>

<ul>
<li>Online search of the NCBI databases directly over a network connection. Pros: you get the latest up-to-date information. Cons: retrieving lots of data over the network can be relatively slow.</li>
<li>Offline search utilises Annokey's local cache of the NCBI databases. Pros: searching the local cache is much faster than online search and does not need a network connection. Cons: the cache data may be incomplete or out of date.</li>
</ul><p>The NCBI provides an entire copy of the gene database in XML format at <a href="ftp.ncbi.nlm.nih.gov">ftp.ncbi.nlm.nih.gov</a>. Annokey can use this XML file to automatically populate its local cache. We also provide a tool called <code>get_ncbi_gene_snapshot_xml.py</code> (described below) for automatically downloading this XML copy of the database.</p>

<h5>
<a name="flexible-search-terms" class="anchor" href="#flexible-search-terms"><span class="octicon octicon-link"></span></a>Flexible search terms</h5>

<p>Search terms can be specified as literal terms or as regular expressions.</p>

<ul>
<li>Literal terms are easy to use but an exact match must be found in the database. Sometimes exact match is too restrictive, for example due to issues with upper- and lower-case letters.</li>
<li>Regular expressions are much more flexible than literal terms, but they require more expertise from the user.</li>
</ul><h6>
<a name="search-results-reported-in-summary-and-detailed-form" class="anchor" href="#search-results-reported-in-summary-and-detailed-form"><span class="octicon octicon-link"></span></a>Search results reported in summary and detailed form</h6>

<p>The search results are provided in summary form, as extra annotations to the input gene list, and also in a more detailed report, as a HTML document. Here is a <a href="http://www.berniepope.id.au/html/annokey_report.html">small example search report</a>.</p>

<ul>
<li>Summary annotations provide a quick overview of the relevance of a key phrase to a gene, and can be incorporated into a workflow using spreadsheets.</li>
<li>The detailed report provides more information than the summary, and contains hyperlinks back to the online NCBI databases for easy reference.</li>
</ul><h3>
<a name="inputs-and-outputs" class="anchor" href="#inputs-and-outputs"><span class="octicon octicon-link"></span></a>Inputs and Outputs</h3>

<p>Annokey's main inputs are:</p>

<ol>
<li>A CSV file representing a set of genes, one gene per row.</li>
<li>A list of search terms sorted in descending order of priority.</li>
</ol><p>Annokey's main outputs are:</p>

<ol>
<li>A CSV file representing an annotated version of the input CSV file. Each row of the CSV file (representing a given gene) is given three extra columns summarising the search results. The extra columns are explained below in an example.</li>
<li>A HTML report file showing more detailed search results for each gene. This includes hyperlinks back to the relevant entries in the NCBI website for easy follow-up investigation.</li>
</ol><h6>
<a name="simple-example-of-inputs-and-outputs" class="anchor" href="#simple-example-of-inputs-and-outputs"><span class="octicon octicon-link"></span></a>Simple example of inputs and outputs</h6>

<p>For example, if your input gene CSV file contains the following rows:</p>

<pre><code>Gene
XRCC2
PALB2
BRCA1
</code></pre>

<p>and your input search term file contains the following lines:</p>

<pre><code>DNA repair
breast cancer
</code></pre>

<p>the output CSV file produced by Annokey will look like this:</p>

<pre><code>Gene,Highest Ranked Match,Matched Terms,Fields
XRCC2,1,DNA repair;breast cancer,Process;GeneRIFs(8/33);Alternative Name
PALB2,1,DNA repair;breast cancer,Process;GeneRIFs(37/78);GeneRIFs(2/78);PMID(2/192);Summary
BRCA1,1,DNA repair;breast cancer,GeneRIFs(28/1275);Description;PMID(4/3698);Process;GeneRIFs(342/1275);Summary;Alternative Name;PMID(19/3698)
</code></pre>

<p>As you can see the output contains a lot of detail about each search term match. Consider the first case for XRCC2:</p>

<pre><code>XRCC2,1,DNA repair;breast cancer,Process;GeneRIFs(8/33);Alternative Name
</code></pre>

<p>In addition to the gene name, the output row contains three extra columns:</p>

<ol>
<li>Highest Ranked Match: the rank of the highest-ranked search term which was found in the search. In this case the rank is 1, which corresponds to the rank of the search term "DNA repair". Search terms are ranked according to their order in the search terms file. The term on the first line is highest ranked (most important), followed by the second line, and so on.</li>
<li>Matched Terms: a list of all the search terms which matched this particular gene. In this case the terms "DNA repair" and "breast cancer" were found.</li>
<li>Fields: the locations in the NCBI gene entry where the search terms were found. In this case the search terms were found in "Process", "GeneRIFs(8/33)" and "Alternative Name".</li>
</ol><h4>
<a name="input-formatting-requirements" class="anchor" href="#input-formatting-requirements"><span class="octicon octicon-link"></span></a>Input formatting requirements</h4>

<h6>
<a name="the-input-csv-file-must-be-formatted-as-follows" class="anchor" href="#the-input-csv-file-must-be-formatted-as-follows"><span class="octicon octicon-link"></span></a>The input CSV file must be formatted as follows:</h6>

<ul>
<li>The file may be comma or tab separated. Annokey looks at the filename extension to guess what format is used (".csv" for comma separated files, ".tsv" for tab separated files) but you may override this with the <code>--delimiter {comma|tab}</code> command line argument.</li>
<li>The first row in the file must be a header row.</li>
<li>Each row may have one or more columns. One of the columns must contain the gene name, and the header for that row must be "Gene".</li>
</ul><p>Here is an example valid gene file in comma separated format:</p>

<pre><code>Sample,Study,Gene,Variant
s100,XYZ,BRCA1,snp
s100,XYZ,XRCC2,indel
s888,ABC,PALB2,snp
s100,ABC,CHECK2,snp
</code></pre>

<p>Notice that there are multiple columns in each row. Annokey only requires the "Gene" column, the other columns are ignored as far as the search is concerned, but are preserved in the output file.</p>

<h6>
<a name="the-search-terms-file-must-be-formatted-as-follows" class="anchor" href="#the-search-terms-file-must-be-formatted-as-follows"><span class="octicon octicon-link"></span></a>The search terms file must be formatted as follows:</h6>

<ul>
<li>Each search term must appear on a separate line.</li>
<li>A search term may be either a literal phrase or a regular expression.</li>
<li>Literal phrases may contain one or more words. Literal phrases are matched exactly by Annokey. For example "DNA" is considered to be different to "dna".</li>
<li>Regular expressions must start with the word "regex" and then be followed by a valid Python regular expression. The <a href="http://docs.python.org/2/howto/regex.html">Python regular expression HOWTO</a> provides a good tutorial on the topic.</li>
<li>Search terms are ranked based on their relative order in the file. The search term on line 1 has rank 1 and so on. Thus the most important search term appears on the first line of the file. The second most important term appears on the second line, and so forth.</li>
</ul><p>A literal search term might look like this:</p>

<pre><code>breast cancer
</code></pre>

<p>A regular expression search term might look like this:</p>

<pre><code>regex [Bb]reast [Cc]ancer(s)?
</code></pre>

<p>This regular expression is more flexible because it will match with upper and lower case characters at the start of each word, and it also allows for an optional "s" at the end of the second word.</p>

<h3>
<a name="installation" class="anchor" href="#installation"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>Annokey currently requires version 2.7 of Python.</p>

<p>The best way to install Annokey is to use the following command:</p>

<pre><code>pip install git+https://github.com/bjpop/annokey.git
</code></pre>

<p>This will automatically download and install the dependencies of Annokey.</p>

<h3>
<a name="online-versus-offline-search" class="anchor" href="#online-versus-offline-search"><span class="octicon octicon-link"></span></a>Online versus offline search</h3>

<p>Annokey obtains its search results by querying the NCBI gene database and the linked PubMed article abstracts. It can access the data in two ways:</p>

<ol>
<li>By sending queries to the online database via the <a href="http://en.wikipedia.org/wiki/Entrez">Entrez</a> programming interface.</li>
<li>By searching in a locally cached copy of the data.</li>
</ol><p>Online queries give you access to the most up-to-date version of the database, but at the cost of being much slower, and requiring an internet connection. The offline version searches in a local copy of the database stored on your computer. Offline search is potentially much faster than online search and does not need an internet access. However, your local copy of the database may not necessarily contain the most up-to-date data.</p>

<p>Annokey automatically keeps a local cache of any data retrieved from an online search. This allows you to incrementally accumulate a local copy of parts of the database, without having to download the entire thing in advance.</p>

<p>By default Annokey keeps its cache of the NCBI Gene Database in a directory called <code>genecache</code>, and a cache of the PubMed article summaries in <code>pubmedcache</code>, however you can specify an alternative locations via the <code>--genecache DIR</code> and <code>--pubmedcache DIR</code> command line arguments. The structure of the cache directories is discussed in the technical details section below.</p>

<h3>
<a name="how-to-use-annokey-from-the-command-line" class="anchor" href="#how-to-use-annokey-from-the-command-line"><span class="octicon octicon-link"></span></a>How to use Annokey from the command line</h3>

<p>You can get help about the Annokey command line by using <code>annokey -h</code> (or <code>--help</code>), which will produce this output:</p>

<pre><code>usage: annokey [-h] [--version] [--online] [--xml FILE] [--organism ORGANISM]
               [--email EMAIL_ADDRESS] [--genecache DIR] [--pubmedcache DIR]
               --terms FILE --genes FILE [--log FILENAME]
               [--delimiter {comma,tab}] [--report FILENAME]

Search NCBI for genes of interest, based on concept-keyword search.

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --online              Search gene information from online (NCBI).
  --xml FILE            Populate gene information from downloaded XML dump of
                        NCBI gene database
  --organism ORGANISM   Name of the organism to search
  --email EMAIL_ADDRESS
                        Your email address. This is required by NCBI for
                        online queries. You do not need to supply an email
                        address for cached queries.
  --genecache DIR       Save a cache of the downloaded results from NCBI gene
                        into this directory
  --pubmedcache DIR     Save a cache of the downloaded results from NCBI
                        pubmed into this directory
  --terms FILE          The tab separated file containing the search terms to
                        be searched.
  --genes FILE          The tab separated file containing the gene information
                        including name of the gene, one gene name per line.
  --log FILENAME        log progress in FILENAME, defaults to annokey_log.txt
  --delimiter {comma,tab}
                        Delimiter for gene file.
  --report FILENAME     Save a detailed search report as HTML page, defaults
                        to annokey_report.html
</code></pre>

<h6>
<a name="required-arguments" class="anchor" href="#required-arguments"><span class="octicon octicon-link"></span></a>Required arguments</h6>

<p>Annokey requires two arguments <code>--terms</code> and <code>--genes</code>, which define the search terms and gene list respectively. The remaining arguments are optional. </p>

<p>Annokey's default mode is to use local offline search in a cached copy of the NCBI Gene and PubMed databases. For example, if the file <code>terms.txt</code> contains a list of search terms and the file <code>genes.csv</code> contains the gene information, then the following command is sufficient to perform a search:</p>

<pre><code>annokey --terms terms.txt --genes genes.csv
</code></pre>

<h6>
<a name="saving-the-annotated-output-to-a-file" class="anchor" href="#saving-the-annotated-output-to-a-file"><span class="octicon octicon-link"></span></a>Saving the annotated output to a file</h6>

<p>Annokey produces a new gene file on standard output. You can save this output to a file using the Unix shell redirection operator "&gt;":</p>

<pre><code>annokey --terms terms.txt --genes genes.csv &gt; annotated_genes.csv
</code></pre>

<h6>
<a name="log-file" class="anchor" href="#log-file"><span class="octicon octicon-link"></span></a>Log file</h6>

<p>As a side-effect, Annokey will produce a log file which contains useful information about the progress of the search. For instance, the log file will record the names of any genes that it could not find in the local cache. By default the log file is written to the file <code>annokey_log.txt</code>, but you can override that with the <code>--log</code> argument:</p>

<pre><code>annokey --terms terms.txt --genes genes.csv --log mylogfile.txt
</code></pre>

<p>The log file also records the command line that was used to execute the program. This can come in handy if you want to run the same command again, but you forgot what the arguments were.</p>

<h6>
<a name="online-search" class="anchor" href="#online-search"><span class="octicon octicon-link"></span></a>Online search</h6>

<p>The effectiveness of offline search depends greatly on the status of your local cache of data.</p>

<p>Alternatively, Annokey can be made to perform an online search with the <code>--online</code> argument. This argument also requires you to specify an email address (as per requirements of the NCBI Gene database):</p>

<pre><code>annokey --online --email foo@bar.com --terms terms.txt --genes genes.csv
</code></pre>

<p>Obviously you should use a real email address instead of <code>foo@bar.com</code>.</p>

<h6>
<a name="organism" class="anchor" href="#organism"><span class="octicon octicon-link"></span></a>Organism</h6>

<p>By default Annokey uses the human version of the NCBI Gene database, but you can specify an alternative organism with the <code>--organism</code> argument:</p>

<pre><code>annokey --organism mouse --terms terms.txt --genes genes.csv
</code></pre>

<h6>
<a name="gene-file-delimiter" class="anchor" href="#gene-file-delimiter"><span class="octicon octicon-link"></span></a>Gene file delimiter</h6>

<p>Multiple columns in the genes input file can be separated by either commas (so-called CSV format) or tabs (so-called TSV format). If the filename ends in ".csv" then Annokey will assume that the file is in CSV format. If the filename ends in ".tsv" then Annokey will assume that the file is in TSV format. However, you can override this behaviour by specifying the desired format using the <code>--delimiter</code> command line argument. The value of <code>--delimiter</code> can be either <code>comma</code> or <code>tab</code>, for example:</p>

<pre><code>annokey --terms terms.txt --genes genes.txt --delimiter comma
</code></pre>

<p>or</p>

<pre><code>annokey --terms terms.txt --genes genes.txt --delimiter tab
</code></pre>

<h6>
<a name="using-a-snapshot-of-the-ncbi-gene-database" class="anchor" href="#using-a-snapshot-of-the-ncbi-gene-database"><span class="octicon octicon-link"></span></a>Using a snapshot of the NCBI gene database</h6>

<p>The NCBI provides a complete snapshot of the gene database which is updated regularly. Annokey can use this snapshot to automatically populate its local cache of the data. In many cases this is the most efficient way to use Annokey. Details about how to obtain a snapshot of the database are provided below.</p>

<pre><code>annokey --terms terms.txt --genes genes.csv --xml Homo_sapiens_20140116041343.xml
</code></pre>

<p>In the example above the snapshot is stored in the file called Homo_sapiens_20140116041343.xml. It stores a record for each gene in the human database.
Annokey will process the XML file and populate the local gene database cache. Future searches can just use the cache directly, and will be substantially faster.</p>

<p>Unfortunately there is no corresponding snapshot of the NCBI article summary database.</p>

<h4>
<a name="populating-the-gene-cache-with-a-complete-download-of-ncbi-gene" class="anchor" href="#populating-the-gene-cache-with-a-complete-download-of-ncbi-gene"><span class="octicon octicon-link"></span></a>Populating the gene cache with a complete download of NCBI gene</h4>

<p>Snapshots of the NCBI gene database for different organisms are provided on the <a href="ftp://ftp.ncbi.nih.gov/gene/DATA/">NCBI ftp site</a>. The snapshot is stored in a compressed binary format which can be converted to XML using a tool called <a href="ftp://ftp.ncbi.nih.gov/asn1-converters/by_program/gene2xml/">gene2xml</a>.</p>

<p>The process of downloading and converting the database snapshot is rather convoluted so we provide a command line tool called <code>get_ncbi_gene_snapshot_xml.py</code> to automate the process. It requires that you have a copy of the <code>gene2xml</code> program in your PATH. You can run the program like so:</p>

<pre><code>get_ncbi_gene_snapshot_xml.py
</code></pre>

<p>the output will look like this:</p>

<pre><code>Starts connecting to server (ftp.ncbi.nlm.nih.gov).
  &gt;&gt; Connecting to server is successful.
Starts getting the last modified time of data.
  &gt;&gt; Getting the last modified time of data is successful.
  &gt;&gt; The last modified time of the data at the server is 20140116041343
Starts downloading data from server.
  &gt;&gt; The file size is 160.39M
  &gt;&gt; Downloading...        100.00% 
  &gt;&gt; Downloading data is successful.
  &gt;&gt; Homo_sapiens_20140116041343.ags.gz is stored.
Starts converting Homo_sapiens_20140116041343.ags.gz to Homo_sapiens_20140116041343.xml.
  &gt;&gt; Converting ...
  &gt;&gt; Converting Homo_sapiens_20140116041343.ags.gz to Homo_sapiens_20140116041343.xml is successful.
Done.
</code></pre>

<p>When it is complete you will have a file called <code>Homo_sapiens_N.xml</code> in your directory, where <code>N</code> is a number representing the date that the snapshot was taken. It should be several gigabytes in size (it was 12 gigabytes on 17 January 2014).</p>

<p>If you have an up-to-date snapshot of the database in the current directory then <code>get_ncbi_gene_snapshot_xml.py</code> will not download a new version:</p>

<pre><code>Starts connecting to server (ftp.ncbi.nlm.nih.gov).
  &gt;&gt; Connecting to server is successful.
Starts getting the last modified time of data.
  &gt;&gt; Getting the last modified time of data is successful.
  &gt;&gt; The last modified time of the data at the server is 20140116041343
You already have the latest version of the file (Homo_sapiens_20140116041343.ags.gz).
You already have the latest version of the xml file (Homo_sapiens_20140116041343.xml).
Done.
</code></pre>

<p>The <code>get_ncbi_gene_snapshot_xml.py</code> program supports several command line arguments. You can view help information by running <code>get_ncbi_gene_snapshot_xml.py -h</code>:</p>

<pre><code>usage: get_ncbi_gene_snapshot_xml.py [-h] [--verbose VERBOSE]
                                     [--ftpsite FTPSITE]
                                     [--filepath GENEFILEPATH] [--force]
                                     [--downloadonly] [--convertonly GENEFILE]
                                     [--gene2xmlpath PROGRAMPATH]

This program downloads gene data from the ftp server and converts data to
xml file by using linux.gene2xml tool. This program checks the last modified
time of the database in the server, when it tries to download the gene
database from the server. If the same version of database is in the local
directory already, this program does not download the database from server. If
you want to download the database anyway, please use the --force option. If
you already have the database file that you need to look up and want to
convert the file, please use the --convertonly option.

optional arguments:
  -h, --help            show this help message and exit
  --verbose VERBOSE     Prints progress messages. The users have to pass the
                        level of messages. 0: No progress messages. 1: Prints
                        what stage the program is on. 2: Prints detail
                        information on progress. 3: Prints callstack when
                        error occurs.
  --ftpsite FTPSITE     The ftp site to be connected for downloading gene
                        database. Default is NCBI server
                        (ftp.ncbi.nlm.nih.gov).
  --filepath GENEFILEPATH
                        The full file path in the ftp site that would be
                        downloaded. The path should start from the root of the
                        ftp site. For example, if you want to download gene
                        database of Archaea from NCBI FTP site, you should
                        pass
                        /gene/DATA/ASN_BINARY/Archaea_Bacteria/Archaea.ags.gz
                        as an argument value. Default is Homo sapiens
                        (/gene/DATA/ASN_BINARY/Mammalia/Homo_sapiens.ags.gz)
  --force               Downloads the latest version of data from the server
                        without regarding the existing file on directory. If
                        there is the same version in local directory, the file
                        would be overwritten.
  --downloadonly        Downloads the latest version of data from the server
                        but not converts the file to xml.
  --convertonly GENEFILE
                        Converts the input file to xml.
  --gene2xmlpath PROGRAMPATH
                        The path where the gene2xml is in. If the path for
                        linux.gene2xml is not in your PATH, you should pass
                        this argument. For example, if linux.gene2xml is in
                        /usr/local/gene2xml/1.3/bin and not in PATH , then you
                        should pass either
                        /usr/local/gene2xml/1.3/bin/linux.gene2xml or
                        /usr/local/gene2xml/1.3/bin.
</code></pre>

<h3>
<a name="technical-details" class="anchor" href="#technical-details"><span class="octicon octicon-link"></span></a>Technical details</h3>

<h6>
<a name="which-parts-of-ncbi-gene-and-pubmed-does-annokey-search" class="anchor" href="#which-parts-of-ncbi-gene-and-pubmed-does-annokey-search"><span class="octicon octicon-link"></span></a>Which parts of NCBI gene and PubMed does Annokey search?</h6>

<p>Annokey searches in the following sections of the NCBI gene database. The diagram below illustrates the parts of the NCBI Gene XML schema which are searched by Annokey. Comments in parenthesis provide additional remarks about how the data is processed and, where relevant, provides the "field names" that Annokey associates with the data:</p>

<pre><code>Entrezgene
    Entrezgene_track-info
        Gene-track/Gene-track_status (we skip entries which are discontinued)
    Entrezgene_gene
        Gene-ref
            Gene-ref_locus (Gene Name)
            Gene-ref_desc (Description)
            Gene-ref_syn
                Gene-ref_syn_E (Synonyms)
    Entrezgene_summary (Summary)
    Entrezgene_comments
        Gene-commentary
            Gene-commentary_type (type of commentary)
            Gene-commentary_heading (heading of commentary)
            Gene-commentary_label (label of commentary)
            Gene-commentary_text (GeneRIFs, if type == generif)
            Gene-commentary_refs
                Pub_pmid/PubMedId (PmIds)
            Gene-commentary_products
            Gene-commentary_comment
                Gene-commentary_text (Pathways, if heading == Pathways)
                Gene-commentary_comment
                    (Interactions, if heading == Interactions)
                    (Conserved Domains, if heading == Conserved Domains)
                    (Function, if label == Function)
                    (Component, if label == Component)
                    (Process, if label == Process)
                    NCBI Reference Sequences (RefSeq) (more commentaries)
        Entrezgene_properties
            Gene-commentary/Gene-commentary_comment
        Entrezgene_prot
            Prot-ref/Prot-ref_name
                Prot-ref_name_E (Alternative Name)
</code></pre>

<p>Having collected all the PubMed IDs from a gene entry (as above) Annokey then searches in the corresponding PubMed database entry. The diagram below illustrates the parts of the PubMed XML schema which are searched by Annokey. The parts marked with an asterisk indicate the fields of text where Annokey searches for key phrases.</p>

<pre><code>PubmedArticle
    MedlineCitation
        Article
            JournalTitle (*)
            ArticleTitle (*)
            Abstract
                AbstractText (*)
</code></pre>

<h6>
<a name="the-structure-of-the-gene-cache-directory" class="anchor" href="#the-structure-of-the-gene-cache-directory"><span class="octicon octicon-link"></span></a>The structure of the gene cache directory</h6>

<p>The local cache of NCBI Gene database entries is structured as follows:</p>

<pre><code>genecache
    organism (e.g. human, mouse)
        hash_dir (a numbered directory in the range [0,255])
            gene_name (e.g. PALB2)
                gene_id (XML file, e.g 79728)
</code></pre>

<p>At the top level is the genecache directory. This has sub-directories for each organism. This has up to 256 sub-directories numbered 0 to 255. The purpose of these sub-directories is to spread out the stored gene information, so that we don't put all the gene files in one sub-directory. Some computer file systems exhibit poor performance when you put too many files in one directory.</p>

<p>The database records for each gene are stored underneath one of the 256 possible hash directories. The exact hash directory is determined by the formula:</p>

<pre><code>md5(gene_name) % 256
</code></pre>

<p>where <code>gene_name</code> is the (string) gene name (such as "PALB2"), md5 is a hash function and % is the modulus operator. We are only using md5 as a way to evenly distribute gene names to directories, so we don't care about its cryptographic properties (particularly its known weaknesses). </p>

<p>For example, the entry for PALB2 is calculated as follows:</p>

<pre><code>md5("PALB2") % 256 = 163
</code></pre>

<p>Which means that PALB2 will be found in genecache/human/163/PALB2/79728. The actual data is stored in the file called 79728, which is in XML format. The number 79728 is the NCBI Gene database ID of PALB2, which is also reflected in the URL to access the entry in the online database <a href="http://www.ncbi.nlm.nih.gov/gene/79728">http://www.ncbi.nlm.nih.gov/gene/79728</a>.</p>

<p>Also note that we normalise the gene name when it is stored in the cache to be all uppercase characters, which means we can reliably find the gene regardless of how it was written in the input gene file or stored in the NCBI database.</p>

<h6>
<a name="the-structure-of-the-pubmed-cache-directory" class="anchor" href="#the-structure-of-the-pubmed-cache-directory"><span class="octicon octicon-link"></span></a>The structure of the PubMed cache directory</h6>

<p>The local cache of PubMed database entries is structured as follows:</p>

<pre><code>pubmedcache
    hash_dir (a numbered directory in the range [0,255])
        PubMed ID (XML file, an integer)
</code></pre>

<p>Each PubMed entry has a unique integer ID. Similar to the gene cache, we spread all entries across 256 sub-directories, using the formula:</p>

<pre><code>md5(PudMed_ID) % 256
</code></pre>

<p>Each PubMed entry is stored in an XML file whose name is equal to its PubMed ID.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Annokey maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
