<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Annokey : Annotation of gene lists with keyword hits from the NCBI gene database and PubMed abstracts" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Annokey</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/bjpop/annokey">View on GitHub</a>

          <h1 id="project_title">Annokey</h1>
          <h2 id="project_tagline">Annotation of gene lists with keyword hits from the NCBI gene database and PubMed abstracts</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/bjpop/annokey/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/bjpop/annokey/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction</h3>

<p>Annokey is a command line tool for annotating gene lists with the results of a keyword (or key phrase) search of the <a href="http://www.ncbi.nlm.nih.gov/gene">NCBI Gene database</a> and linked <a href="http://www.ncbi.nlm.nih.gov/pubmed">PubMed</a> article abstracts. Its purpose is to help users prioritise genes  by relevance to a domain of interest, such as "breast cancer" or "DNA repair" etc. The user steers the search by specifying a ranked list of keywords and terms that are likely to be highly correlated with their domain of interest.</p>

<h3>
<a name="features" class="anchor" href="#features"><span class="octicon octicon-link"></span></a>Features</h3>

<p>Annokey's main features are:</p>

<h6>
<a name="online-and-offline-search" class="anchor" href="#online-and-offline-search"><span class="octicon octicon-link"></span></a>Online and offline search</h6>

<ul>
<li>Online search access the NCBI databases directly over a network connection. Pros: you get the latest, must up-to-date information. Cons: retrieving lots of data over the network can be relatively slow.</li>
<li>Offline search utilises Annokey's local cache of the NCBI databases. Pros: searching the local cache is much faster than online search, and does not need a network connection. Cons: the cache data may be incomplete or not up-to-date.</li>
</ul><p>The NCBI provides an entire copy of the gene database in XML format at <a href="ftp.ncbi.nlm.nih.gov">ftp.ncbi.nlm.nih.gov</a>. Annokey can use this XML file to automatically populate its local cache. We also provide a tool called <code>genetoxml.py</code> for automatically downloading this XML copy of the database.</p>

<h5>
<a name="flexible-search-terms" class="anchor" href="#flexible-search-terms"><span class="octicon octicon-link"></span></a>Flexible search terms</h5>

<p>Search terms can be specified as literal terms or as regular expressions.</p>

<ul>
<li>Literal terms are easy to use but an exact match must be found in the database. Sometimes exact match is too restrictive, for example due to issues with upper and lower case letters.</li>
<li>Regular expressions are much more flexible than literal terms, but they require more expertise from the user.</li>
</ul><h6>
<a name="search-results-reported-in-summary-and-detailed-form" class="anchor" href="#search-results-reported-in-summary-and-detailed-form"><span class="octicon octicon-link"></span></a>Search results reported in summary and detailed form</h6>

<p>The search results are provided in summary form, as extra annotations to the input gene list, and also in a more detailed report, as a HTML document.</p>

<ul>
<li>Summary annotations provide a quick overview of the relevance of a key phrase to a gene, and can be incorporated into a workflow using spreadsheets.</li>
<li>The detailed report provides more information than the summary, and contains hyperlinks back to the online NCBI databases for easy reference.</li>
</ul><h3>
<a name="inputs-and-outputs" class="anchor" href="#inputs-and-outputs"><span class="octicon octicon-link"></span></a>Inputs and Outputs</h3>

<p>Annokey's main inputs are:</p>

<ol>
<li>A CSV file representing a set of genes, one gene per row.</li>
<li>A list of search terms sorted in descending order of priority.</li>
</ol><p>Annokey produces two main outputs:</p>

<ol>
<li>A CSV file representing an annotated version of the input CSV file. Each row of the CSV file (representing a given gene) is given three extra columns summarising the search results. The extra columns are explained below in an example.</li>
<li>A HTML report file showing more detailed search results for each gene. This includes hyper-links back to the relevant entries in the NCBI website for easy follow-up investigation.</li>
</ol><h6>
<a name="simple-example-of-inputs-and-outputs" class="anchor" href="#simple-example-of-inputs-and-outputs"><span class="octicon octicon-link"></span></a>Simple example of inputs and outputs</h6>

<p>For example, if your input gene CSV file contains the following rows:</p>

<pre><code>Gene
XRCC2
PALB2
BRCA1
</code></pre>

<p>and your input search term file contains the following lines:</p>

<pre><code>DNA repair
breast cancer
</code></pre>

<p>the output CSV file produced by Annokey will look like this:</p>

<pre><code>Gene,Highest Ranked Match,Matched Terms,Fields
XRCC2,1,DNA repair;breast cancer,Process;GeneRIFs(8/33);Alternative Name
PALB2,1,DNA repair;breast cancer,Process;GeneRIFs(37/78);GeneRIFs(2/78);PMID(2/192);Summary
BRCA1,1,DNA repair;breast cancer,GeneRIFs(28/1275);Description;PMID(4/3698);Process;GeneRIFs(342/1275);Summary;Alternative Name;PMID(19/3698)
</code></pre>

<p>As you can see the output contains a lot of detail about each search term match. Consider the first case for XRCC2:</p>

<pre><code>XRCC2,1,DNA repair;breast cancer,Process;GeneRIFs(8/33);Alternative Name
</code></pre>

<p>In addition to the gene name, the output row contains three extra columns:</p>

<ol>
<li>Highest Ranked Match: the rank of the highest-ranked search term which was found in the search. In this case the rank is 1, which corresponds to the rank of the search term "DNA repair". Search terms are ranked according to their order in the search terms file. The term on the first line is highest ranked (most important), followed by the second line, and so on.</li>
<li>Matched Terms: a list of all the search terms which matched this particular gene. In this case the terms "DNA repair" and "breast cancer" were found.</li>
<li>Fields: the locations in the NCBI gene entry where the search terms were found. In this case the search terms were found in "Process", "GeneRIFs(8/33)" and "Alternative Name".</li>
</ol><h4>
<a name="input-formatting-requirements" class="anchor" href="#input-formatting-requirements"><span class="octicon octicon-link"></span></a>Input formatting requirements</h4>

<h6>
<a name="the-input-csv-file-must-be-formatted-as-follows" class="anchor" href="#the-input-csv-file-must-be-formatted-as-follows"><span class="octicon octicon-link"></span></a>The input CSV file must be formatted as follows:</h6>

<ul>
<li>The file may be comma or tab separated. By default Annokey assumes the file will be comma separated, but you may override this with the <code>--delimiter tab</code> command line argument.</li>
<li>The first row in the file must be a header row.</li>
<li>Each row may have one or more columns. One of the columns must contain the gene name, and the header for that row must be "Gene".</li>
</ul><p>Here is an example valid gene file in comma separated format:</p>

<pre><code>Sample,Study,Gene,Variant
s100,XYZ,BRCA1,snp
s100,XYZ,XRCC2,indel
s888,ABC,PALB2,snp
s100,ABC,CHECK2,snp
</code></pre>

<p>Notice that there are multiple columns in each row. The only column that Annokey considers is the one with the header "Gene".</p>

<p>Annokey will preserve the other columns in its output, and add its own annoations at the end of each row.</p>

<h6>
<a name="the-search-term-file-must-be-formatted-as-follows" class="anchor" href="#the-search-term-file-must-be-formatted-as-follows"><span class="octicon octicon-link"></span></a>The search term file must be formatted as follows:</h6>

<ul>
<li>Each search term must appear on a separate line.</li>
<li>A search term may be either a literal phrase or a regular expression.</li>
<li>Literal phrases may contain one or more words. Literal phrases are matched exactly by Annokey. This includes character case. For example "DNA" is considered to be different to "dna".</li>
<li>Regular expressions must start with the word "regex" and then be followed by a valid Python regular expression. The <a href="http://docs.python.org/2/howto/regex.html">Python regular expression HOWTO</a> provides a good tutorial on the topic.</li>
<li>Search terms are ranked based on their relative order in the file. The search term on line 1 has rank 1 and so on. Thus the most important search term appears on the first line of the file. The second most important term appears on the second line, and so forth.</li>
</ul><p>A literal search term might look like this:</p>

<pre><code>breast cancer
</code></pre>

<p>A regular expression search term might look like this:</p>

<pre><code>regex [Bb]reast [Cc]ancer(s)?
</code></pre>

<p>This regular expression is more flexible than the previous literal term because the regular expression will match with upper and lower case characters at the start of each work, and it also allows for an optional "s" at the end of the second word.</p>

<h3>
<a name="installation" class="anchor" href="#installation"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>Annokey currently requires version 2.7 of Python.</p>

<p>The best way to install Annokey is to use the following command:</p>

<pre><code>pip install git+https://github.com/bjpop/annokey.git
</code></pre>

<p>This will automatically download and install the dependencies of Annokey.</p>

<h3>
<a name="online-versus-offline-search" class="anchor" href="#online-versus-offline-search"><span class="octicon octicon-link"></span></a>Online versus offline search</h3>

<p>Annokey obtains its search results by querying the NCBI gene database. It can access that data in two ways:</p>

<ol>
<li>By sending queries to the online database via the Entrez programming interface.</li>
<li>By searching in a locally cached copy of the data.</li>
</ol><p>Online queries give you access to the most up-to-date version of the database, but at the cost of being much slower, and requiring an internet connection. The offline version searches in a local copy of the database stored on your computer. Offline search is potentially much faster than online search and does not need an internet access. However, your local copy of the database may not necessarily contain the most up-to-date data.</p>

<p>Annokey automatically keeps a local cache of any data retrieved from an online search. This allows you to incrementally accumulate a local copy of parts of the database, without having to download the entire thing in advance.</p>

<p>By default Annokey keeps its cache of the NCBI Gene Database in a directory called <code>genecache</code>, however you can specify an alternative location via a command line argument. The structure of the cache directory is as follows:</p>

<pre><code>genecache
    organism (e.g. human, mouse)
        hash_dir (a numbered directory in the range [0,255])
            gene_name (e.g. PALB2)
                gene_id (XML file, e.g 79728)
</code></pre>

<p>At the top level is the genecache directory. This has sub-directories for each organism. This has up to 256 sub-directories numbered 0 to 255. The purpose of these sub-directories is to spread out the stored gene information, so that we don't put all the gene files in one sub-directory. Some computer file systems exhibit poor performance when you put too many files in one directory.</p>

<hr><p><strong>Technical note</strong>: the database records for each gene are stored underneath one of the 256 possible hash directories. The exact hash directory is determined by the formula:</p>

<pre><code>md5(gene_name) % 256
</code></pre>

<p>where <code>gene_name</code> is the (string) gene name (such as "PALB2"), md5 is a hash function and % is the modulus operator. We are only using md5 as a way to evenly distribute gene names to directories, so we don't care about its cryptographic properties (particularly its known weaknesses). </p>

<p>For example, the entry for PALB2 is calculated as follows:</p>

<pre><code>md5("PALB2") % 256 = 163
</code></pre>

<p>Which means that PALB2 will be found in genecache/human/163/PALB2/79728. The actual data is stored in the file called 79728, which is in XML format. The number 79728 is the NCBI Gene database ID of the PALB2, which is also reflected in the URL to access the entry in the online database <a href="http://www.ncbi.nlm.nih.gov/gene/79728">http://www.ncbi.nlm.nih.gov/gene/79728</a>.</p>

<p>Also note that we normalise the gene name when it is stored in the cache to be all uppercase characters, which means we can reliably find the gene regardless of how it was written in the input gene file or stored in the NCBI database.</p>

<hr><h3>
<a name="how-to-use-annokey-from-the-command-line" class="anchor" href="#how-to-use-annokey-from-the-command-line"><span class="octicon octicon-link"></span></a>How to use Annokey from the command line</h3>

<p>You can get help about the Annokey command line by using <code>annokey -h</code> (or <code>--help</code>), which will produce this output:</p>

<pre><code>usage: annokey [-h] [--version] [--online] [--organism ORGANISM]
               [--email EMAIL_ADDRESS] [--genecache DIR] [--pubmedcache DIR]
               --keys FILE --genes FILE [--log FILENAME]
               [--delimiter {comma,tab}]

Search NCBI for genes of interest, based on concept-keyword search.

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --online              Search gene information from online (NCBI).
  --organism ORGANISM   Name of the organism to search
  --email EMAIL_ADDRESS
                        Your email address. This is required byNCBI for online
                        queries. You do not needto supply an email address for
                        cached queries.
  --genecache DIR       Save a cache of the downloaded results from NCBI gene
                        into this directory
  --pubmedcache DIR     Save a cache of the downloaded results from NCBI
                        pubmed into this directory
  --keys FILE           The tab separated file containing the keywords to be
                        searched.
  --genes FILE          The tab separated file containing the gene information
                        including name of the gene, one gene name per line.
  --log FILENAME        log progress in FILENAME, defaults to annokey_log.txt
  --delimiter {comma,tab}
                        delimiter for gene file
</code></pre>

<p>Annokey requires two arguments <code>--keys</code> and <code>--genes</code>, which define the search terms and gene list respectively. The remaining arguments are optional. </p>

<p>Annokey's default mode is to use local "offline" search in a cached copy of the NCBI Gene database. For example, if the file <code>keys.txt</code> contains a list of search terms (one term per line in free text format) and the file <code>genes.txt</code> contains the gene information (one gene per line in CSV format), then the following command is sufficient to perform a search:</p>

<pre><code>annokey --keys keys.txt --genes genes.csv
</code></pre>

<p>Annokey produces a new gene file on standard output, which you will normally want to redirect to a file:</p>

<pre><code>annokey --keys keys.txt --genes genes.csv &gt; annotated_genes.csv
</code></pre>

<p>As a side-effect, Annokey will produce a log file which contains useful information about the progress of the search. By default this is written to the file <code>annokey_log.txt</code>, but you can override that with the <code>--log</code> argument:</p>

<pre><code>annokey --keys keys.txt --genes genes.csv --log mylogfile.txt &gt; annotated_genes.csv
</code></pre>

<p>One of the more useful aspects of the log file is that it records the command line that was used to execute the program. This can come in handy if you want to run the same command again, but you forgot what the arguments were.</p>

<p>The effectiveness of offline search depends greatly on the status of your local cache of the NCBI gene database.</p>

<p>Alternatively, Annokey can be made to perform an online search with the <code>--online</code> argument. This argument also requires you to specify an email address (as per requirements of the NCBI Gene database):</p>

<pre><code>annokey --online --email foo@bar.com --keys keys.txt --genes genes.csv --log mylogfile.txt &gt; annotated_genes.csv
</code></pre>

<p>Obviously you should use a real email address instead of <code>foo@bar.com</code>.</p>

<p>By default Annokey uses the human version of the NCBI Gene database, but you can specify an alternative organism with the <code>--organism</code> argument:</p>

<pre><code>annokey --organism mouse --online --email foo@bar.com --keys keys.txt --genes genes.csv --log mylogfile.txt &gt; annotated_genes.csv
</code></pre>

<h3>
<a name="required-format-of-input-files" class="anchor" href="#required-format-of-input-files"><span class="octicon octicon-link"></span></a>Required format of input files</h3>

<p>The input file specifying the search terms is in free text format. Each line of the file specifies a search term. Annokey is case sensitive, so the following terms are considered different:</p>

<pre><code>dna repair
DNA repair
DNA Repair
dNA rEpAiR
</code></pre>

<p>The input file specifying the list of genes is in Comma Separated Format (CSV). Each line of the file specifies an individual gene. There must be one column in the file which specifies the name of the gene. The columns of the file must have headings, and the gene column must be entitled <code>Gene</code>. The file may have zero or more other columns. Additional columns are ignored by Annokey, but will be preserved in the output of the program.</p>

<h4>
<a name="format-of-the-input-genes-file" class="anchor" href="#format-of-the-input-genes-file"><span class="octicon octicon-link"></span></a>Format of the input genes file</h4>

<p>The <code>--genes</code> command line argument specifies the name of the input genes file. The format of the file is quite flexible, but a few constraints must be satisfied. The first row of the file must be a heading row, and exactly one of the headings must be the word 'Gene'. If there are multiple columns in the file then they can be separated by either commas (so-called CSV format) or tabs (so-called TSV format). If the filename ends in ".csv" then Annokey will assume that the file is in CSV format. If the filename ends in ".tsv" then Annokey will assume that the file is in TSV format. However, you can override this behaviour by specifying the desired format using the <code>--delimiter</code> command line argument. The value of <code>--delimiter</code> can be either <code>comma</code> or <code>tab</code>, for example:</p>

<pre><code>annokey --keys keys.txt --genes mygenes.txt --delimiter comma
</code></pre>

<p>or</p>

<pre><code>annokey --keys keys.txt --genes mygenes.txt --delimiter tab
</code></pre>

<p>The relative order of the columns in the genes file does not matter, as long as one of them contains the names of genes, and that same column has "Gene" as its heading. Annokey will ignore the other columns, but it will preserve them in the output, in their same relative position.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Annokey maintained by <a href="https://github.com/bjpop">bjpop</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
